crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -72,13 +133,110 @@ where
     }
 
     let extra_scripts: Vec<ScriptHash> = provided_scripts
+        .iter()
+        .map(ScriptHash::from)
+        .collect::<BTreeSet<_>>()
         .difference(&required_scripts)
         .cloned()
         .collect();
+
     if !extra_scripts.is_empty() {
         return Err(InvalidScripts::ExtraneousScriptWitnesses(extra_scripts));
     }
 
+    let required_script_inputs = resolved_inputs
+        .iter()
+        .filter_map(|input_output| {
+            input_output
+                .1
+                .address()
+                .ok()
+                .and_then(|address| match address {
+                    amaru_kernel::Address::Shelley(shelley_address) => {
+                        if shelley_address.payment().is_script() {
+                            if let Some(script) = provided_scripts
+                                .iter()
+                                .find(|script| &script.hash == shelley_address.payment().as_hash())
+                            {
+                                return Some((input_output, script));
+                            }
+                        }
+                        None
+                    }
+                    amaru_kernel::Address::Byron(_) | amaru_kernel::Address::Stake(_) => None,
+                })
+        })
+        .collect::<Vec<_>>();
+
+    let mut input_datum_hashes: BTreeSet<Hash<32>> = BTreeSet::new();
+    let mut inputs_missing_datum: Vec<&TransactionInput> = Vec::new();
+
+    required_script_inputs
+        .into_iter()
+        .for_each(|((input, output), script)| match output.has_datum() {
+            None => {
+                if !(matches!(script.script, PseudoScript::PlutusV3Script(..))
+                    || matches!(script.script, PseudoScript::NativeScript(..)))
```

Comment:
Dangerous; we might forget to update it later with PlutusV4; I would suggest using a complete pattern match here.

Reply:
I insist <3 

And it's a rule of thumb for pretty much everything; avoid wildcards and partial pattern match whenever possible. It makes the code harder to maintain in the long run and it's only usually one line shorter not to.